tidb_exporter/main.go

```diff
@@ -0,0 +1,144 @@
+package main
+
+import (
+	"database/sql"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+// Exporter collects database query stats and exports them using
+// the prometheus metrics package.
+type Exporter struct {
+	db       *sql.DB
+	tidbOpts tidbOpts
+}
+
+type tidbOpts struct {
+	address  string
+	username string
+	password string
+}
+
+// NewExporter returns an initialized Exporter.
+func NewExporter(opts tidbOpts) (*Exporter, error) {
+	db, err := accessDatabase(opts.username, opts.password, opts.address, dbname)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &Exporter{
+		db:       db,
+		tidbOpts: opts,
+	}, nil
+}
+
+// Describe describes all the metrics ever exported by tidb_exporter. It
+// implements prometheus.Collector.
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	ch <- queryErrorDesc
+}
+
+// Collect fetches database query stats and delivers them
+// as Prometheus metrics. It implements prometheus.Collector.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	var queryError float64
+	label, err := probeQuery(e.db)
+	if err != nil {
+		queryError = 1
+	}
+	ch <- prometheus.MustNewConstMetric(
+		queryErrorDesc, prometheus.GaugeValue, queryError, e.tidbOpts.address, label,
+	)
+}
+
+func checkParameters(opts tidbOpts) {
+	if opts.address == "" {
+		log.Fatalf("missing startup parameter: --tidb.address")
+	}
+
+	if opts.username == "" {
+		log.Fatalf("missing startup parameter: --tidb.username")
+	}
+
+	if opts.password == "" {
```

Comment:
tidb's password can't be empty?

Reply:
In fact, checking opts.password is optional. I add it for security. I' ll remove it in next pr.